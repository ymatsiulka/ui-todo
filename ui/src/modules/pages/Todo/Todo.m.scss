@import 'styles/variables.scss';

.todoCheckboxItem {
    display: flex;
    justify-content: space-between;
    padding: 10px;
    padding: 8px;
    border-bottom: 1px solid #E3E3E3;
    -webkit-transition: all 0.1s cubic-bezier(0.165, 0.84, 0.44, 1);
    transition: all 0.1s cubic-bezier(0.165, 0.84, 0.44, 1);

    &:hover {
        box-shadow: 0 3px 3px rgba(0, 0, 0, 0.05)
    }
}

.wrapper {
    position: relative;

    &::before {
        background-image: linear-gradient(to left, rgba(113, 165, 251, 0.8) 0%, rgba(184, 97, 238, 0.75) 100%), url('assets/mountains.jpg');
        content: '';
        background-size: cover;
        background-position: center;
        position: absolute;
        top: 0;
        width: 100%;
        height: 300px;
        z-index: -1;
    }

    .content {
        padding: 20px;
        position: relative;
        z-index: 1;
    }
}

.todoContent {
    display: flex;
    flex-direction: column;
    height: 100%;
    width: 400px;
    margin: 0 auto;
    margin-top: 50px;
}

.todoListContainer {
    width: 400px;
    height: 400px;
    background-color: $white;
    margin: 0 auto;
    margin-top: 10px;
    border-radius: 2px;
    overflow-y: auto;
}