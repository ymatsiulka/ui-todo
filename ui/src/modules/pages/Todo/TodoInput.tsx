import React from 'react';
import { Input } from 'modules/shared';
import { input } from 'appConstants';
// import styles from './TodoInput.m.scss';

interface TodoInputProps {
  value: string;
  onChangeHandler: React.ChangeEventHandler<HTMLInputElement>;
  onKeyUpHandler: React.KeyboardEventHandler<HTMLInputElement>;
}

const TodoInput: React.FC<TodoInputProps> = ({ value, onChangeHandler, onKeyUpHandler }) => {
  return (
    <Input
      value={value}
      id="todo-input"
      className="todo-input"
      placeholder=""
      name={'todo-input'}
      label={''}
      title={'todo-input'}
      type={input.text}
      onChange={onChangeHandler}
      onKeyUp={onKeyUpHandler}
    >
      <span className="check__box" style={{ marginLeft: '0px', marginRight: '0px' }} />
    </Input>
  );
};

export default TodoInput;
