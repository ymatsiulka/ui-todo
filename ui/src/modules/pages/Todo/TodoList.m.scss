.todoList {
    width: 400px;
    height: 400px;
    background-color: white;
    margin: 0 auto;
    margin-top: 10px;
    border-radius: 2px;
    overflow-y: auto;
}