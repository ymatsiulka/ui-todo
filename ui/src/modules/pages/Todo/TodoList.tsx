import React from 'react';
import { TodoResponse } from '../../../types/api';
import Input from '../../shared/input/Input';
import { input } from 'appConstants';

interface TodoListProps {
  todos: TodoResponse[];
}

const TodoList: React.FC<TodoListProps> = ({ todos }) => (
  <ul className="todo-list-container">
    {todos.map((t: TodoResponse) => {
      return (
        <li id={t.id.toString()} data-value={t.id.toString()}>
          <Input
            name={t.id.toString()}
            label={t.name}
            type={input.checkbox}
            checked={t.isCompleted}
            onChange={(e) => {}}
          />
        </li>
      );
    })}
  </ul>
);

export default TodoList;
