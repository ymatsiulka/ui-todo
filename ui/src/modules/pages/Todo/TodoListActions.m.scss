.todoListActions {
    display: flex;
    justify-content: space-between;
    align-items: center;
    background-color: white;
    padding: 10px;
    border-bottom-left-radius: 2px;
    border-bottom-right-radius: 2px;
    box-shadow: 0 3px 3px rgba(0, 0, 0, 0.05)
}

.todoListActions>.left>span {
    margin: 0;
    background-color: white;
    color: grey;
    padding: 2px 2px;

}

.todoListActions>.middle>button {
    margin-left: 5px;
    margin-right: 5px;
}