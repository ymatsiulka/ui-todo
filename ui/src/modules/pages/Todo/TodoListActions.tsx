import React from 'react';
import { TodoPageStatuses } from '../../../types/frontend';
import Button from '../../../modules/shared/button/Button';

interface TodoListActionsProps {
  todoPageStatus: TodoPageStatuses;
  uncompletedItemsCount: number;
  onClickAllHandler: React.MouseEventHandler<HTMLButtonElement>;
  onClickActiveHandler: React.MouseEventHandler<HTMLButtonElement>;
  onClickCompletedHandler: React.MouseEventHandler<HTMLButtonElement>;
  onClickClearCompletedHandler: React.MouseEventHandler<HTMLButtonElement>;
}

const TodoListActions: React.FC<TodoListActionsProps> = ({
  todoPageStatus,
  uncompletedItemsCount,
  onClickAllHandler,
  onClickActiveHandler,
  onClickCompletedHandler,
  onClickClearCompletedHandler,
}) => {
  const isActiveButton = (isActive: boolean) => `fs-12 ${isActive && 'c-blue'}`;

  return (
    <div className="todo-list-actions">
      <div className="left">
        <span className="fs-12">{uncompletedItemsCount} items left </span>
      </div>
      <div className="middle">
        <Button
          className={isActiveButton(todoPageStatus == TodoPageStatuses.All)}
          title="All"
          onClick={onClickAllHandler}
        />
        <Button
          className={isActiveButton(todoPageStatus == TodoPageStatuses.Active)}
          title="Active"
          onClick={onClickActiveHandler}
        />
        <Button
          className={isActiveButton(todoPageStatus == TodoPageStatuses.Completed)}
          title="Completed"
          onClick={onClickCompletedHandler}
        />
      </div>
      <div className="right">
        <Button className="fs-12" title="Clear Completed" onClick={onClickClearCompletedHandler} />
      </div>
    </div>
  );
};

export default TodoListActions;
