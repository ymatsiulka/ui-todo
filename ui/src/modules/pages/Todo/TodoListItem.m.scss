.todoCheckboxItem {
    display: flex;
    justify-content: space-between;
    border-bottom: 1px solid #E3E3E3;
    -webkit-transition: all 0.1s cubic-bezier(0.165, 0.84, 0.44, 1);
    transition: all 0.1s cubic-bezier(0.165, 0.84, 0.44, 1);
    padding: 10px;

    &:hover {
        box-shadow: 0 3px 3px rgba(0, 0, 0, 0.05)
    }
}
