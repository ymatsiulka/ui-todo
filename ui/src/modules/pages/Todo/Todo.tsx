import React, { useRef, useState } from 'react';
import { Spacer, Typography } from 'modules/shared';
import { TodoPageStatuses } from 'types/frontend';
import { keys } from 'appConstants';
import TodoListActions from './TodoListActions';
import { useAppDispatch, useAppSelector } from 'hooks';
import { addTodo, clearTodos, moveTodos, uncompletedSelectItemsCount } from 'store/features/todos/todosSlice';
import styles from './Todo.m.scss';
import TodoList from './TodoList';
import TodoInput from './TodoInput';

const Todo: React.FC = () => {
  const [todoPageStatus, setTodoPageStatus] = useState<TodoPageStatuses>(TodoPageStatuses.All);
  const [todoName, setTodoName] = useState('');
  const dragTodo = useRef<number>(0);
  const draggedOverTodo = useRef<number>(0);
  const onDragEndHandler = (): void => {
    dispatch(moveTodos({ firstIndex: dragTodo.current, secondIndex: draggedOverTodo.current }));
  };
  const onDragStartHandler = (index: number): void => {
    dragTodo.current = index;
  };
  const onDragEnterHandler = (index: number): void => {
    draggedOverTodo.current = index;
  };
  const todosState = useAppSelector((state) => state.todos);
  const todosItems = todosState.items;
  const dispatch = useAppDispatch();

  const onClickCompletedHandler = (): void => {
    setTodoPageStatus(TodoPageStatuses.Completed);
  };

  const onClickActiveHandler = (): void => {
    setTodoPageStatus(TodoPageStatuses.Active);
  };

  const onClickAllHandler = (): void => {
    setTodoPageStatus(TodoPageStatuses.All);
  };

  const onChangeHandler = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setTodoName(e.target.value);
  };

  const onClickClearCompletedHandler = (): void => {
    dispatch(clearTodos());
  };

  const onKeyUpHandler = (e: React.KeyboardEvent<HTMLInputElement>): void => {
    if (e.key === keys.enter) {
      setTodoName('');
      dispatch(addTodo({ todoName }));
    }
  };

  const uncompletedItemsCount = uncompletedSelectItemsCount(todosState);
  return (
    <div className={styles.content}>
      <div className={styles.todoContent}>
        <Typography variant="h1" className="c-white" text="T O D O" />
        <Spacer top={15} bottom={15} />
        <TodoInput value={todoName} onChangeHandler={onChangeHandler} onKeyUpHandler={onKeyUpHandler} />
        <Spacer top={10} bottom={10} />
        <TodoList
          todoItems={todosItems}
          todoPageStatus={todoPageStatus}
          onDragStartHandler={onDragStartHandler}
          onDragEndHandler={onDragEndHandler}
          onDragEnterHandler={onDragEnterHandler}
        />
        <TodoListActions
          onClickAllHandler={onClickAllHandler}
          onClickActiveHandler={onClickActiveHandler}
          onClickCompletedHandler={onClickCompletedHandler}
          onClickClearCompletedHandler={onClickClearCompletedHandler}
          uncompletedItemsCount={uncompletedItemsCount}
          todoPageStatus={todoPageStatus}
        />
        <footer>{todosItems.length > 1 && 'Drag and drop to reorder list'}</footer>
      </div>
      <div className="background-image" />
    </div>
  );
};

export default Todo;
