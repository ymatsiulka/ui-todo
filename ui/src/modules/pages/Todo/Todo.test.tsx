import React, { useState } from 'react';
import { fireEvent, render } from '@testing-library/react';
import { useAppDispatch, useAppSelector } from 'hooks';
import { type TodoState } from 'types/redux/TodoState';
import { TodoPageStatuses } from 'types/frontend';
import Todo from './Todo';
import { keys } from 'appConstants';

const mockDispatch = jest.fn();
const mockSetTodoName = jest.fn();
const mockSetTodoPageStatus = jest.fn();
const mockUseState = useState as jest.Mock;
const mockUseAppDispatch = useAppDispatch as jest.Mock;
const mockUseAppSelector = useAppSelector as jest.Mock;

const state: TodoState = {
  items: [
    {
      id: 1,
      order: 1,
      name: 'name1',
      isCompleted: false,
    },
    {
      id: 2,
      order: 2,
      name: 'name2',
      isCompleted: false,
    },
  ],
  nextId: 2,
};

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useEffect: jest.fn((x) => x()),
}));

jest.mock('hooks', () => ({
  useAppDispatch: jest.fn(),
  useAppSelector: jest.fn(),
}));

jest.mock('modules/shared/organisms/TodoList', () =>
  jest.fn(() => (
    <ul>
      <li>some text</li>
    </ul>
  )),
);

jest.mock('modules/shared/moleculas/TodoInput', () =>
  jest.fn(({ onChangeHandler, onKeyUpHandler }) => (
    <input data-testid="input" type="text" onChange={onChangeHandler} onKeyUp={onKeyUpHandler} />
  )),
);

jest.mock('modules/shared/organisms/TodoListActions', () =>
  jest.fn(({ onClickAllHandler, onClickActiveHandler, onClickCompletedHandler, onClickClearCompletedHandler }) => (
    <div>
      <button data-testid="all-button" onClick={onClickAllHandler} />
      <button data-testid="active-button" onClick={onClickActiveHandler} />
      <button data-testid="complete-button" onClick={onClickCompletedHandler} />
      <button data-testid="clear-button" onClick={onClickClearCompletedHandler} />
    </div>
  )),
);

beforeEach(() => {
  mockDispatch.mockClear();
  mockSetTodoName.mockClear();
  mockUseState.mockImplementationOnce(() => ['', mockSetTodoName]);
  mockSetTodoPageStatus.mockClear();
  mockUseState.mockImplementationOnce(() => [TodoPageStatuses.All, mockSetTodoPageStatus]);
  mockUseAppSelector.mockReturnValueOnce(state);
  mockUseAppDispatch.mockReturnValueOnce(mockDispatch);
});

test('Todo should render properly', async () => {
  // Arrange
  // Act
  const { asFragment } = render(<Todo />);

  // Assert
  expect(asFragment()).toMatchSnapshot();
});

test('Todo when no items render properly', async () => {
  // Arrange
  mockUseAppSelector.mockReturnValueOnce({
    items: [],
    nextId: 0,
  });

  // Act
  const { asFragment } = render(<Todo />);

  // Assert
  expect(asFragment()).toMatchSnapshot();
});

test('Todo when all button clicked should update state', async () => {
  // Arrange
  // Act
  const { getByTestId } = render(<Todo />);

  const allButton = getByTestId('all-button');
  fireEvent.click(allButton);

  // Assert
  expect(mockSetTodoPageStatus).toHaveBeenCalledTimes(1);
  expect(mockSetTodoPageStatus).toHaveBeenCalledWith(TodoPageStatuses.All);
});

test('Todo when active button clicked should update state', async () => {
  // Arrange
  // Act
  const { getByTestId } = render(<Todo />);

  const activeButton = getByTestId('active-button');
  fireEvent.click(activeButton);

  // Assert
  expect(mockSetTodoPageStatus).toHaveBeenCalledTimes(1);
  expect(mockSetTodoPageStatus).toHaveBeenCalledWith(TodoPageStatuses.Active);
});

test('Todo when complete button clicked should update state', async () => {
  // Arrange
  // Act
  const { getByTestId } = render(<Todo />);

  const completeButton = getByTestId('complete-button');
  fireEvent.click(completeButton);

  // Assert
  expect(mockSetTodoPageStatus).toHaveBeenCalledTimes(1);
  expect(mockSetTodoPageStatus).toHaveBeenCalledWith(TodoPageStatuses.Completed);
});

test('Todo when clear button clicked should dispatch store', async () => {
  // Arrange
  // Act
  const { getByTestId } = render(<Todo />);

  const clearButton = getByTestId('clear-button');
  fireEvent.click(clearButton);

  // Assert
  expect(mockDispatch).toHaveBeenCalledTimes(1);
});

test('Todo when input changed should update state', async () => {
  // Arrange
  // Act
  const { getByTestId } = render(<Todo />);

  const input = getByTestId('input');
  fireEvent.change(input, { target: { value: 'some text' } });

  // Assert
  expect(mockSetTodoName).toHaveBeenCalledTimes(1);
  expect(mockSetTodoName).toHaveBeenCalledWith('some text');
});

test('Todo when enter key up should dispatch store', async () => {
  // Arrange
  // Act
  const { getByTestId } = render(<Todo />);

  const input = getByTestId('input');
  fireEvent.keyUp(input, { key: keys.enter });

  // Assert
  expect(mockDispatch).toHaveBeenCalledTimes(1);
  expect(mockSetTodoName).toHaveBeenCalledTimes(1);
  expect(mockSetTodoName).toHaveBeenCalledWith('');
});

test('Todo when some key up should do nothing', async () => {
  // Arrange
  // Act
  const { getByTestId } = render(<Todo />);

  const input = getByTestId('input');
  fireEvent.keyUp(input, { key: 'F' });

  // Assert
  expect(mockDispatch).toHaveBeenCalledTimes(0);
  expect(mockSetTodoName).toHaveBeenCalledTimes(0);
});
