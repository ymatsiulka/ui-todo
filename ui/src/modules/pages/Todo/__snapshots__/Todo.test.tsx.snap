// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo should render properly 1`] = `
<DocumentFragment>
  <div
    class="content"
  >
    <div
      class="todoContent"
    >
      <div
        bottom="10"
        style="margin-top: 10px; margin-bottom: 10px;"
        top="10"
      />
      <h1
        class="c-white"
      >
        T O D O
      </h1>
      <div
        bottom="15"
        style="margin-top: 15px; margin-bottom: 15px;"
        top="15"
      />
      <input
        data-testid="input"
        type="text"
      />
      <div
        bottom="10"
        style="margin-top: 10px; margin-bottom: 10px;"
        top="10"
      />
      <ul>
        <li>
          some text
        </li>
      </ul>
      <div>
        <button
          data-testid="all-button"
        />
        <button
          data-testid="active-button"
        />
        <button
          data-testid="complete-button"
        />
        <button
          data-testid="clear-button"
        />
      </div>
      <div
        bottom="10"
        style="margin-top: 10px; margin-bottom: 10px;"
        top="10"
      />
      <footer
        id="dnd-footer"
      >
        Drag and drop to reorder list
      </footer>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Todo when no items render properly 1`] = `
<DocumentFragment>
  <div
    class="content"
  >
    <div
      class="todoContent"
    >
      <div
        bottom="10"
        style="margin-top: 10px; margin-bottom: 10px;"
        top="10"
      />
      <h1
        class="c-white"
      >
        T O D O
      </h1>
      <div
        bottom="15"
        style="margin-top: 15px; margin-bottom: 15px;"
        top="15"
      />
      <input
        data-testid="input"
        type="text"
      />
      <div
        bottom="10"
        style="margin-top: 10px; margin-bottom: 10px;"
        top="10"
      />
      <ul>
        <li>
          some text
        </li>
      </ul>
      <div>
        <button
          data-testid="all-button"
        />
        <button
          data-testid="active-button"
        />
        <button
          data-testid="complete-button"
        />
        <button
          data-testid="clear-button"
        />
      </div>
      <div
        bottom="10"
        style="margin-top: 10px; margin-bottom: 10px;"
        top="10"
      />
      <footer
        id="dnd-footer"
      >
        Drag and drop to reorder list
      </footer>
    </div>
  </div>
</DocumentFragment>
`;
